File: structured_streaming-subject/src/main/scala/structured_streaming/data_pretreatment/Pretreatment.java
package structured_streaming.data_pretreatment;

import org.apache.commons.lang.StringUtils;

import static structured_streaming.data_pretreatment.DataPretreatment.getDataStr;

import java.io.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Classname Pretreatment
 * Date 2020/7/17 14:53
 * Created by LanKorment
 * 预处理收集到的Telemetry数据，转换成可以使用spark解析的格式
 */
public class Pretreatment {
    public static void main(String[] args) throws Exception {
        StringBuffer str = getDataStr("D:\\WorkSpace\\Telemetry\\dump-interface.txt");
        pretreatment(data2str(str));
    }

    public static String[] data2str(StringBuffer strbuff) {
        String data = strbuff.toString();
        //System.out.println(data);
        String[] result = data.split("------- ");
        return result;
    }

    public static String replaceBlank(String str) {
        String dest = "";
        if (str != null) {
            Pattern p = Pattern.compile("\\s*|\t|\r|\n");
            Matcher m = p.matcher(str);
            dest = m.replaceAll("");
        }
        return dest;
    }

    public static void pretreatment(String[] strarr) throws Exception {

        for (int i = 1; i < strarr.length; i++) {

            String str = strarr[i].replace("-------\n", "");
            String header = StringUtils.substringBetween(str, "2020-", "]") + "]";
            String body = StringUtils.substringAfter(str, header);

            /*
            //将Header加入到Json中
            String data = replaceBlank(body);
            String json = new StringBuffer(data).insert(1,"\"Header\":" + "\"" + header + "\"" + ",").toString();*/

            //未将Header加入到body中
            String json = replaceBlank(body);

            String filePath="D:\\WorkSpace\\Telemetry\\test.txt";
            //文件追加内容
            FileOutputStream fos = new FileOutputStream(filePath,true);
            fos.write(json.getBytes());
            fos.write("\n".getBytes());

            //System.out.println(header);
            //System.out.println(json);
        }
    }
}

File: structured_streaming-subject/src/main/scala/structured_streaming/parse/Pretreatment.scala
package structured_streaming.parse

import java.io.{File, FileOutputStream, FileWriter}
import java.util.regex.{Matcher, Pattern}

import org.apache.commons.lang.StringUtils
import structured_streaming.data_pretreatment.DataPretreatment.getDataStr

/**
  * Classname Pretreatment
  * Description TODO
  * Date 2020/7/21 10:53
  * Created by LanKorment
  * 使用scala来预处理数据
  */
object Pretreatment {
  def main(args: Array[String]): Unit = {
    preprocessingData("D:\\WorkSpace\\Telemetry\\dump-cpu.txt","D:\\WorkSpace\\Telemetry\\test.txt");
  }

  //传入参数原始数据的路径以及经过处理后的数据的输出路径
  def preprocessingData(inputpath : String,outputpath : String): Unit ={
    val str = getDataStr(inputpath)
    pretreatment(data2str(str),outputpath)
  }

  def data2str(strbuff: StringBuffer): Array[String] = {
    val data = strbuff.toString
    //System.out.println(data);
    val result = data.split("------- ")
    result
  }

  def replaceBlank(str: String): String = {
    var dest = ""
    if (str != null) {
      val p = Pattern.compile("\\s*|\t|\r|\n")
      val m = p.matcher(str)
      dest = m.replaceAll("")
    }
    dest
  }

  //strarr为将文件转换成的字符串，即data2str()的返回，传入参数path为预处理数据输出的位置
  @throws[Exception]
  def pretreatment(strarr: Array[String],outpath : String ): Unit = {
    var i = 1
    while ( {
      i < strarr.length
    }) {
      val str = strarr(i).replace("-------\n", "")
      val header = StringUtils.substringBetween(str, "2020-", "]") + "]"
      val body = StringUtils.substringAfter(str, header)

      //将Header加入到Json中
      /*val data = replaceBlank(body);
      val json = new StringBuffer(data).insert(1,"\"Header\":" + "\"" + header + "\"" + ",").toString();*/

      //未将Header加入到body中
      val json = replaceBlank(body)
      val filePath = outpath
      //文件追加内容
      val fos = new FileOutputStream(filePath, true)
      fos.write(json.getBytes)
      fos.write("\n".getBytes)
      //System.out.println(header);
      //System.out.println(json);

      {
        i += 1; i - 1
      }
    }
  }

}
